# Limpieza de datos
Crearemos los archivos necesarios para poder realizar los análisis correspondientes.

In [1]:
import pandas as pd

## Consumo de Productos Derivados del Petróleo

En la fila 7 se encuentran los títulos de las columnas. Con datos hasta la fila 300.

En la columna A se encuentra la fecha, en la columna L se encuentra la gasolina superior, en la K la gasolina regular, en la G el diesel bajo azufre y el gas propano en la I.

In [2]:
df = pd.read_excel('data/raw/CONSUMO.xlsx', header=6, nrows=294, usecols='A,G,I,K,L')

In [3]:
df.tail()

Fecha  Diesel bajo azufre  Gas licuado de petróleo  Gasolina regular  \
288 2024-01-01          1359012.49                548124.45         830708.13   
289 2024-02-01          1340174.42                526897.85         818740.16   
290 2024-03-01          1393324.52                523990.91         870771.70   
291 2024-04-01          1428143.44                531880.19         847353.15   
292 2024-05-01          1401052.37                536754.38         894533.14   

     Gasolina superior  
288          658083.66  
289          654059.60  
290          671997.05  
291          633520.57  
292          692427.94

Renombraremos las columnas para facilitar su manipulación.

In [4]:
df.columns = ['Fecha', 'Diesel', 'Propano', 'Regular', 'Super']
df.head()

Fecha  Diesel        Propano    Regular      Super
0 2000-01-01     0.0  194410.476190  202645.20  308156.82
1 2000-02-01     0.0  174710.552381  205530.96  307766.31
2 2000-03-01     0.0  189234.066667  229499.56  331910.29
3 2000-04-01     0.0  174330.607143  210680.40  315648.08
4 2000-05-01     0.0  191745.147619  208164.34  319667.97

Convertimos a una serie de tiempo

In [5]:
df['Fecha'] = pd.to_datetime(df['Fecha'])
df.set_index('Fecha', inplace=True)
df.sort_index(inplace=True)

In [6]:
df.tail()

Diesel    Propano    Regular      Super
Fecha                                                  
2024-01-01  1359012.49  548124.45  830708.13  658083.66
2024-02-01  1340174.42  526897.85  818740.16  654059.60
2024-03-01  1393324.52  523990.91  870771.70  671997.05
2024-04-01  1428143.44  531880.19  847353.15  633520.57
2024-05-01  1401052.37  536754.38  894533.14  692427.94

## Importación de Productos Derivados del Petróleo

En la fila 7 se encuentran los títulos de las columnas. Con datos hasta la fila 288.

En la columna A se encuentra la fecha, en la columna L se encuentra la gasolina superior, en la K la gasolina regular, en la G el diesel bajo azufre y el gas propano en la I.

In [7]:
df = pd.read_excel('data/raw/IMPORTACION.xlsx', header=6, nrows=281, usecols='A,G,I,K,L')

In [8]:
df.tail()

Fecha  Diesel bajo azufre  Gas licuado de petróleo  Gasolina regular  \
276 2024-01-01          1409097.15                701570.80         914133.32   
277 2024-02-01          1236861.75                916541.70         740662.25   
278 2024-03-01          1477038.00                675157.48         838270.93   
279 2024-04-01          1294706.12                473940.66         886132.77   
280 2024-05-01          1470870.09                684864.46         939656.18   

     Gasolina superior  
276          712333.33  
277          650360.11  
278          620077.74  
279          687017.96  
280          696970.30

Renombraremos las columnas para facilitar su manipulación.

In [9]:
df.columns = ['Fecha', 'Diesel', 'Propano', 'Regular', 'Super']
df.head()

Fecha  Diesel        Propano    Regular      Super
0 2001-01-01     0.0  194065.738095  177776.50  373963.96
1 2001-02-01     0.0  170703.380952  123115.99  243091.07
2 2001-03-01     0.0  161837.371429  161726.42  312084.38
3 2001-04-01     0.0  163048.642857  127338.74  285054.89
4 2001-05-01     0.0  171518.861905  168730.19  300913.67

Convertimos a una serie de tiempo

In [10]:
df['Fecha'] = pd.to_datetime(df['Fecha'])
df.set_index('Fecha', inplace=True)
df.sort_index(inplace=True)

In [11]:
df.tail()

Diesel    Propano    Regular      Super
Fecha                                                  
2024-01-01  1409097.15  701570.80  914133.32  712333.33
2024-02-01  1236861.75  916541.70  740662.25  650360.11
2024-03-01  1477038.00  675157.48  838270.93  620077.74
2024-04-01  1294706.12  473940.66  886132.77  687017.96
2024-05-01  1470870.09  684864.46  939656.18  696970.30

## Precios de los Productos Derivados del Petróleo

Un formato no muy amigable, se leerá cada producto por separado. Luego, se unirán en un solo DataFrame.

### Gasolina Superior

In [12]:
df_super = pd.read_excel('data/raw/PRECIOS.xlsx', header=22, nrows=12, usecols='A:X')

In [13]:
df_regular = pd.read_excel('data/raw/PRECIOs.xlsx', header=43, nrows=12, usecols='A:X')

In [14]:
df_diesel = pd.read_excel('data/raw/PRECIOs.xlsx', sheet_name=1, header=1, nrows=12, usecols='A:X')

In [15]:
df_diesel.head()

MES/AÑO  2002   2003   2004   2005   2006   2007   2008   2009   2010  ...  \
0    ENERO  8.95  10.46  12.19  15.68  19.59  18.68  26.02  19.11  24.13  ...   
1  FEBRERO  8.66  11.50  11.90  15.10  19.16  17.64  26.09  17.97  23.59  ...   
2    MARZO  8.69  12.83  12.17  16.01  18.51  18.80  27.84  16.73  23.96  ...   
3    ABRIL  9.38  12.22  13.71  17.89  19.35  20.34  30.47  16.60  24.56  ...   
4     MAYO  9.44  10.63  13.79  18.31  20.38  20.91  32.31  17.33  24.42  ...   

    2015   2016   2017   2018   2019   2020   2021   2022   2023   2024  
0  18.86  13.65  18.35  20.74  19.87  21.09  18.02  26.40  35.79  29.16  
1  19.81  13.02  17.89  20.39  21.09  18.74  19.46  29.29  33.02  30.72  
2  20.18  14.80  17.60  20.07  21.86  17.11  21.21  36.21  30.67  29.97  
3  20.04  15.12  17.60  21.22  21.94  14.91  20.65  32.95  29.78  29.78  
4  21.32  16.26  16.86  22.29  21.97  14.09  21.75  34.83  27.43  28.17  

[5 rows x 24 columns]

Necesitamos una función que coloque el dataframe en el formato correcto. 

In [16]:
df_diesel.at[1, 2002]

8.66

In [17]:
def format_df(df: pd.DataFrame, product: str) -> pd.DataFrame:
    columns = ['Fecha', product]
    data = []

    for year in df.columns:
        for i in range(1, 13):  # Representa los meses
            # Omitir la primera columna
            if year == 'MES/AÑO':
                continue

            year = int(year)
            timestamp = pd.Timestamp(f'{year}-{i}-01')
            price = df.at[i-1, year]
            data.append({'Fecha': timestamp, product: price})

    # Crear el DataFrame usando la lista de diccionarios
    df_temp = pd.DataFrame(data, columns=columns)
    return df_temp

In [18]:
new_df_super = format_df(df_super, 'Super')

In [19]:
new_df_regular = format_df(df_regular, 'Regular')

In [20]:
new_df_diesel = format_df(df_diesel, 'Diesel')

Unimos los DataFrames

In [21]:
df = pd.merge(new_df_super, new_df_regular, on='Fecha')
df = pd.merge(df, new_df_diesel, on='Fecha')

In [22]:
df.head()

Fecha  Super  Regular  Diesel
0 2002-01-01  12.55    12.19    8.95
1 2002-02-01  12.29    11.90    8.66
2 2002-03-01  12.58    12.17    8.69
3 2002-04-01  14.11    13.71    9.38
4 2002-05-01  14.16    13.79    9.44